### ISSUE TITLE
Wrong type in last evt in s12df_to_s12l

### ISSUE THREAD

**Issue by gonzaponte** [next-exp/IC#146]
The function is supposed to return a dictionary with dictionaries holding two np.ndarrays. When looping over the full input DataFrame the last event is not cast to np.ndarray and remains as a list.

**Comment by jacg** [comment]
Could you provide a test that demonstrates the problem?

**Comment by gonzaponte** [comment]
Done in my fork gonzaponte/issue146.

**Comment by jacg** [comment]
In Python we usually do not care about the concrete types of objects. Can you reframe this test in terms of some *capabiltiy* of the objects? In other words, as it stands, this test looks too much like it is inspired by a failure to understand Duck Typing. And we wouldn't want future readers of our code to get the impression that we don't understand Duck Typing.

**Comment by gonzaponte** [comment]
The problem is that the difference of capabilities of a duck and a person are too many. I want to have full functionality of np arrays, so the easiest way is to check it behaves as a np array is to ask for its type, isn’t it?

**Comment by jacg** [comment]
But you don't want the full functionality of numpy arrays; you only need that functionality which you are actually using.

Why don't you start with *one* bit of numpy functionality that is missing from whatever is being returned right now, that is stopping you from achieving your broader goals.

Besides, I don't know why we are talking about numpy arrays, when the test fails because it demands a list in a situation where it gets a generator. This in itself is a good illustration why you should write tests in terms of capabilities rather than types! And these capabilities should be ones you *actually* need, not ones you think you need, because they happen to be provided by the type you happen to be thinking of.

Hint: you will find it almost impossible to persuade me to accept code containing a type check demanding a list (or any other contrete type or hierarchy). Sequence-like behaviour? OK; iterator-like behaviour? OK; broadcasting? OK; the concrete list type? NO WAY!

Checks for concrete types have no place in Python programming (except when talking to low-level components). The sooner you accept this, the happier your Python experience (and that of other users of your code) will be.

**Comment by gonzaponte** [comment]
It is true the test fails there, although that was just a secondary check (and that’s why we’re talking about np arrays: the primary check that is currently implemented within the same test).

When JJ first wrote that function it was supossed to return {evt_idx: {pk_idx: [np.array, np.array] } }. For the last event, however, it returns  {evt_idx: {pk_idx: (list, list) } }. So

1.- For consistency I find annoying that the last event has a tuple instead of a list. I don’t care much about this because, as you pointed out, the functionality is the same. Do you think consistency is not that important to change this?

2.- I can easily check one bit of numpy (e.g. check that it can be summed with .sum()), but I find this insufficient as it does not mean it behaves like a numpy array in other situations (ask for .max(), for instance). These objects (the output of the function) are meant for analysis, so full functionality is probably necessary, don’t you think?

**Comment by jacg** [comment]
> It is true the test fails there, although that was just a secondary check (and that’s why we’re talking about np arrays: the primary check that is currently implemented within the same test).

We're talking about why your test should check duck types rather than concrete types, and the fact that the test is failing *before* we even get to the numpy array part *as a consequence of a concrete rather than duck type check* is a perfect demonstration of *why* tests should check for duck types rather than contrete types:

+ The test fails (at this point) because it is written badly.
+ The test is written badly because it rejects a perfectly correct object.
+ The test rejects a correct object because it checks for a concrete type rather than a duck type. 
+ Conclusion: check for duct types, not concrete types.

1. The only consistency that matters (in Python) is consistency of *behaviour*. The concrete types *really do not matter*. This is a cornerstone of Python programming. Python is beautifully polymorphic ... until someone with misguided ideas about consistency of concrete types comes along and breaks it. (Don't worry, the developers of Python themselves have fallen into this trap, so you're in good company. But don't take that as licence to put in type checks: they have removed them in Python3)

1. Check for a small number of behaviours that are most commonly and imediately expected of the object that appears in that place. For now, take YAGNI as your guiding principle on this one.

> The problem is that the difference of capabilities of a duck and a person are too many.

That rather depends on what you need to do with them, doesn't it? If you want something that is warm-blooded, or something that can provide a DNA sample, or something that has two eyes, then both a duck and a person will do just fine.

Duck typing really is a fundamental part of programming in Python with the least pain. Embrace it, and both your and your colleagues will be happier and more productive for it.

I realize that 'used for analysis' makes it difficult to imagine a concrete type other than numpy array that could provide the required capabilites, but we shouldn't make a habit of letting our lack of imagination lead us to developing bad habits. Test for capability rather that concrete type.

**Comment by gonzaponte** [comment]
Ok. A couple of comments:

- Regardless of the discussion, the test was failing because I was doing it
wrong. My bad.

- I agree with the principle (100%), however I find it convenient (in this
case) to combine all tests into a single one. It is simpler and clearer (in
my opinion). Just as we don't take pep8 as a religion, but as a guideline.

Anyway... The proposed change (for the function) is in my fork. I will
rewrite the test in terms of duck capabilities. Should I write a single
test checking all capabilities or one test per capability?

**Comment by jacg** [comment]
I have edited out your huge copypasta. Are you aware that both the 'emails' you receive and your replies are actually comments in the conversation on the GitHub issue you opened?

This is not the place to explain why top-posting is an abominable practice. Top-posting may be tolerable in some circumstances, but in a medium like this (GitHub conversations) it is much, much worse than in email.

Please do not top-post in GitHub conversations. Resist your email agent's attempts to make you top-post in GitHub conversations.

On to the meat of the matter. You seem to think that I am taking issue with you having placed two asserts in the single test. Yes, it is true that that is less than ideal, and you yourself have pointed out why it is a problem. But compared to the *real* issue, it is so insignificant, that I haven't been addressing this at all! Really: I've been saying 'the test fails where it should pass because it checks a concrete type, rather than a duck type', but the message you have been receiving seems to have been 'you have to split the test so that the incorrect failure doesn't stop you from seeing the second part'.

Let's first solve the first-order problem of expressing the test in terms of duck types. Once we're happy with that, we can think about the minor question of whether or not to split the test into two. Remember, *none* of these experiments and trials will end up in `nextic/master`, they will be squashed out. So you don't have to get the whole thing completely right in one go.

> Should I write a single test checking all capabilities or one test per capability?

Start with one test. Let's be sure that we are happy with how duck type checks do the job for us. Once we're happy with the duck type checks, then we can think about how best to split them up into separate tests.

And please do not try to write behavioural tests for every single numpy array feature you can think of! Remember the law of diminishing returns: every extra test you add to this group, will be less valuable to you than the previous one. The first one is the most important. I would stop after one or two or three.

**Comment by jacg** [comment]
Please avoid using the term *cast* in Python. Call it *conversion*. Yes, I know you'll find lots of people in cyberspace calling it casting: we don't have to repeat their mistakes.

While the meaning of the terms *casting*, *conversion* and *coercion* may change across contexts or communities, in many of those it shouldn't be too controversial to say that something like the following meanings apply

+ cast: even though the type system claims that an object has type `S` treat it as if it had type `T`
+ convert: create a *new* object of type `T` from an object of type `S`

In short, *in Python* use the word 'convert' do describe the process of taking one object and using it as a source of information in the process of creating an object of a different type.

(Yes, the `cast` function in `configure.py` is on my hit-list ... it's just got very low priority.)

**Comment by gonzaponte** [comment]
1. Indeed, I wasn't aware of my replies being in "email format". Sorry about that, I will edit them.
2. I wasn't talking about the multiple-asserts-in-a-single-test topic. If you took that from my first bullet point in the previous comment, I was referring to the fact that the assert statement was checking something that ***could not be*** correct and therefore it was failing regardless of the concrete vs duck types discussion (it is still a good example why duck types must be ones used in tests). I should write tests to check the tests...
3. Where does this "cast vs convert discussion" comes from?
4. Is it correct to use the hasattr built-in function to check duck types?

**Comment by gonzaponte** [comment]
Ok, I found where the cast discussion comes from and I have fixed it.
I have pushed a new version of the test checking duck types (using hasattr). Can you take a look at it?

**Comment by jacg** [comment]
> Where does this "cast vs convert discussion" comes from?

The function called `cast_to_arrays` that you add in your branch that bears the name of the issue we are discussing.

Eeek! I now see that, in that function, you are modifying an existing object rather than returning a new one. Hmmm. Sorry, I don't have time now. This will have to wait.

**Comment by gonzaponte** [comment]
Ok, I've changed that.

**Comment by gonzaponte** [comment]
This issue may no longer matter as the relevant part of the code is being revisited because of issue #149. However, it can still be used to increase the number of tests over the pmaps_functions module.

**Comment by jacg** [comment]
Hmm, concerning the top-posting. I notice that JJ top-posts his replies in email all the time, but GitHub is clever enough to hide them in an expandable section that looks approximately like this: `...`. So, somehow GitHub failed to do that with your top-post. If you top-post in a way that GitHub manages to hide (like JJ), then feel free top-post all you want.

**Comment by gonzaponte** [comment]
I have rebased onto master and extended one of the tests to check duck types. I think this concludes this issue. I will open a PR.

**Comment by jjgomezcadenas** [comment]
I think this thread is solved
